RavenDB-22955 Replication - Usability: Add Replication Progress

shiranshalom opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22955/Replication-Usability-Add-Replication-Progress

### Additional description

* Add `LastAcceptedChangeVectorFromDestination`, `SourceDatabaseChangeVector`, `LastSentEtag`, `FromToString` to ‘/databases/*/tasks’ & ‘/databases/*/task’ EPs
* Create new EP [_'/databases/*/replication/progress'_] to get replication ongoing tasks progress

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [x] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

shiranshalom committed:
RavenDB-21922 - Add LastAcceptedChangeVectorFromDestination, SourceDatabaseChangeVector, LastSentEtag, FromToString to ‘/databases/*/tasks’ & ‘/databases/*/task’ EPs

shiranshalom committed:
RavenDB-21922 - implement new EP for replication progress ["/databases/*/replication/progress"]

# Conflicts:
#	src/Raven.Server/Documents/Replication/ReplicationLoader.cs

shiranshalom committed:
RavenDB-22955 - Add AverageProcessedPerSecond & Implement new EP for Internal Replication progress

garayx commented on src/Raven.Server/Documents/Handlers/Processors/Replication/ReplicationHandlerProcessorForGetOngoingTasksProgress.cs:
```diff
@@ -0,0 +1,78 @@
+
+                var taskId = node.TaskId;
+
+                if (names.Count > 0 && names.Contains(node.Name) == false)
```
can we simply return the empty dictionary if `names.Count == 0` ?

shiranshalom replied:
no, because `names` is not a required parameter.. 
if `names.Count == 0`, we still need to collect all the existing stats

garayx commented on src/Raven.Server/Documents/Handlers/Processors/Replication/ReplicationHandlerProcessorForGetOngoingTasksProgress.cs:
```diff
@@ -0,0 +1,78 @@
+        protected override Task HandleRemoteNodeAsync(ProxyCommand<ReplicationTaskProgress[]> command, OperationCancelToken token) => RequestHandler.ExecuteRemoteAsync(command, token.Token);
+
+        public IDictionary<long, ReplicationTaskProgress> GetProcessesProgress(DocumentsOperationContext context, StringValues names)
+        {
```
shouldn't we handle here `if names == default` case ?

shiranshalom replied:
I don't think so.. if `names == default` it's fine, we will collect all :)

garayx commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -1896,6 +1901,66 @@ public static bool IsOfTypePreventDeletions(ReplicationBatchItem item)
+                AverageProcessedPerSecond = handler.Metrics.GetProcessedPerSecondRate() ?? 0.0
+            };
+
+            var collections = Database.DocumentsStorage.GetCollections(documentsContext).Select(x => x.Name).ToList();
```
can we drop the `ToList()`?

```suggestion
            var collections = Database.DocumentsStorage.GetCollections(documentsContext).Select(x => x.Name);
```

shiranshalom replied:
definitely! done

garayx commented on src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:
```diff
@@ -95,7 +95,7 @@ internal Table EnsureRevisionTableCreated(Transaction tx, CollectionName collect
             var tableName = collection.GetTableName(CollectionTableType.Revisions);
 
-            if (_tableCreated.Contains(collection.Name) == false)
+            if (tx.IsWriteTransaction && _tableCreated.Contains(collection.Name) == false)
```
why do we change that ? it doesn't seem to be related to this PR

shiranshalom replied:
I noticed this issue while adding the `GetNumberOfRevisionsToProcess` and `GetNumberOfRevisionTombstones` methods. it’s not directly related to this PR, but it was causing issues in certain scenarios so I fixed it

garayx commented on src/Raven.Server/Documents/Revisions/RevisionsStorage.cs:
```diff
@@ -2743,5 +2743,31 @@ public long GetNumberOfRevisionDocuments(DocumentsOperationContext context)
+                return 0;
+            }
+
+            var table = EnsureRevisionTableCreated(context.Transaction.InnerTransaction, collectionName, RevisionsSchema);
```
cannot you use this ? so no need to change the method
```
            var table = new Table(RevisionsSchema, context.Transaction.InnerTransaction);

```

shiranshalom replied:
done. but I wonder if we still need to keep the change in `EnsureRevisionTableCreated`, what do you think?

garayx commented on src/Raven.Server/Documents/Commands/Replication/GetReplicationOngoingTasksProgressCommand.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Raven.Server.Documents.Commands.Replication
+{
+    internal sealed class GetReplicationOngoingTasksProgressCommand : RavenCommand<ReplicationTaskProgress[]>
```
why don't we want to expose it to client api ?

shiranshalom replied:
TBH, I saw that `GetEtlTaskProgressCommand` is on the server, so I thought this should be handled the same way

shiranshalom committed:
RavenDB-22955 - fixing PR comments#1

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 commented:
@shiranshalom 
1.  I can see replication progress is exposed for PullReplicationAsHub, however I can't find anchor to match that info with data from tasks endpoint. 

Example: create hub with 2 sinks. Then each sink contains progress and only task name (which is always the same)

2. in replication progress I can see totals of TotalNumberOfAttachmentTombstones and TotalNumberOfRevisionTombstones, but I can't see NumberOfAttachmentTombstonesToProcesss, NumberOfRevisionTombstonesToProcess. Do we want to expose and show that ?
3. Please expose last  Database ETag from db stats in tasks endpoint next to OngoingTaskReplication.LastSentEtag

I rebased this PR on latest 6.2 and waiting for your changes in order to continue.

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

shiranshalom committed:
RavenDB-22955 - add lastDatabaseEtag to OngoingTaskReplication

# Conflicts:
#	src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs

shiranshalom commented:
@ml054 
> 
> 1. I can see replication progress is exposed for PullReplicationAsHub, however I can't find anchor to match that info with data from tasks endpoint.
> 
> Example: create hub with 2 sinks. Then each sink contains progress and only task name (which is always the same)
> 
`ProcessesProgress` should have 2 results (one for each sink), perhaps we can leverage `FromToString` to differentiate them by extracting the database name and URL?
![GetReplicationProgressEP2-1Hub2Sinks-2](https://github.com/user-attachments/assets/4c467c07-141e-4bd8-83a3-fd03c8845196)

> 2. in replication progress I can see totals of TotalNumberOfAttachmentTombstones and TotalNumberOfRevisionTombstones, but I can't see NumberOfAttachmentTombstonesToProcesss, NumberOfRevisionTombstonesToProcess. Do we want to expose and show that ?

I wasn’t able to extract `NumberOfAttachmentTombstonesToProcess` and `NumberOfRevisionTombstonesToProcess` separately, but they are included in `NumberOfDocumentTombstonesToProcess`, so we could use that as a combined measure in this case

> 3. Please expose last  Database ETag from db stats in tasks endpoint next to OngoingTaskReplication.LastSentEtag
> 

done :)

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

shiranshalom committed:
RavenDB-22955 - add progress for internal replication for sharded databases & expose fields for OngoingTaskPullReplicationAsHub

# Conflicts:
#	src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs
#	test/SlowTests/Server/Documents/OngoingTasks/ReplicationOngoingTaskProgressTests.cs

shiranshalom committed:
RavenDB-22955 - Add DestinationNodeTag for InternalReplicationTaskProgress & Adjust tests

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

shiranshalom committed:
RavenDB-22955 - exposed errors for external & pull as sink replication tasks

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress - internal

shiranshalom commented:
@ml054 I’ve exposed the latest errors (if any exist) for the `replication/tasks` EP, as follows:
* **External replication:** If there is no existing `OutgoingHandler`, will attempt to retrieve the last error for the task.
* **Pull replication as sink:** If the connection status is `NotActive`, will attempt to retrieve the last error for the task.

_Note:_
* I didn’t implement this for pull replication as hub because the sink is the side that initiates the connection. In case of failure (e.g., the sink is down), there’s no connection on the hub side, and therefore no task details.
* I didn’t implement this for internal replication either, simply because there isn’t an ongoing task for it. However:
In my opinion, adding it to the `/outgoing-internal-replication/progress` EP would miss the purpose of this EP.
Perhaps for internal replication, if there’s no progress, we could direct the user to the Ongoing Tasks Stats view, where exceptions should be displayed, or alternatively, refer them to the Admin Logs?
If we choose to add it to the progress endpoint, it might be better to ensure it behaves the same way for other tasks for consistency. Let me know what you think :)

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

# Conflicts:
#	src/Raven.Studio/.storybook/preview.js

# Conflicts:
#	src/Raven.Studio/.storybook/storeDecorator.tsx

ml054 committed:
RavenDB-22955 fixed linter warnings

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 commented:
I will have to rebase that on latest v6.2

ml054 commented:
Rebased and pushed.

mateuszbartosik committed:
RavenDB-22955 Styling adjustments for replication progress

mateuszbartosik commented:
@ml054 styling adjustments have been pushed

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress

ml054 commented:
@shiranshalom I think we are ready. Feel free to convert this PR from draft to 'regular'

ppekrol commented:
@shiranshalom can you check failing tests? 'AuthenticationDebugPackageTests' are related.

shiranshalom committed:
RavenDB-22955 - fix failing tests

shiranshalom commented:
> @shiranshalom can you check failing tests? 'AuthenticationDebugPackageTests' are related.

@ppekrol Done! I’ve tested the recent changes for internal, external, and hub replication, and everything looks great. I think we’re ready to go! 😊

garayx commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -1898,6 +1913,68 @@ public static bool IsOfTypePreventDeletions(ReplicationBatchItem item)
+            var overallDuration = Stopwatch.StartNew();
+
+            foreach (var collection in collections)
+            {
```
is this is right for filtered replication?

shiranshalom replied:
You're right. I believe it's still valid since we process all collections but skip sending those that don’t match `AllowedHubToSinkPaths` or `AllowedSinkToHubPaths`. Maybe it's worth adding some information about this in the Studio? @ml054  what do you think?

garayx commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -271,6 +272,137 @@ public static void WriteEtlTaskProgress<TWriter>(this TWriter writer, JsonOperat
+        public static void WriteReplicationTaskProgress<TWriter>(this TWriter writer, JsonOperationContext context, IEnumerable<IReplicationTaskProgress> progress)
+        where TWriter : IBlittableJsonTextWriter
+        {
+            writer.WriteStartObject();
```
I wonder if all this can be done by implementing `ToJson()` on `IReplicationTaskProgress` or `taskStats.ProcessesProgress`?

Similar to `EtlHandlerProcessorForDebugStats`

ml054 committed:
RavenDB-22955 Replication - Usability: Add Replication Progress - sink progress

ppekrol committed:
RavenDB-22955 renames
